SQL: `explain (debug) select 3` crashes server

eg:
```
[sql --host=127.0.0.1 --pgport=44101 --certs=/tmp/cli-test561717278 -e explain (debug) select 3]
panic: runtime error: slice bounds out of range

goroutine 277 [running]:
github.com/cockroachdb/cockroach/sql.(*selectNode).Next(0xc820742000, 0xc82010b9c0)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/select.go:122 +0x3c0
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func1(0xece418154, 0xba13b48, 0x2162c80, 0xc820529101, 0x28)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/executor.go:422 +0x308
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func2(0xc82059cdc0, 0x1)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/executor.go:466 +0x121
github.com/cockroachdb/cockroach/client.(*Txn).exec(0xc82059cdc0, 0xc8201e03c8, 0x14)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/client/txn.go:405 +0x2b2
github.com/cockroachdb/cockroach/client.(*DB).Txn(0xc820332a50, 0xc8201e03c8, 0x7f2946ac3400)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/client/db.go:463 +0x15b
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc820332a50, 0x7f2948305228, 0xc8205290e0, 0xc8206be000, 0x0,
 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/executor.go:463 +0x600
github.com/cockroachdb/cockroach/sql.(*Executor).execStmts(0xc820332a50, 0xc820740000, 0x19, 0xc8206be000, 0x0, 0x0, 0x0
, 0x0, 0x0, 0x0)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/executor.go:318 +0x33d
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc820332a50, 0xc8205cc00d, 0x4, 0x0, 0x0, 0x0, 0x0, 
0x0, 0x0, 0x0, ...)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/executor.go:266 +0x5e5
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc8201781a0, 0xc820740000, 0x19, 0x0, 0x0, 0x0,
 0x0, 0x0, 0x0, 0x7f2944146001, ...)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:550 +0x1fb
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc8201781a0, 0xc8201781d8, 0x0, 0x0)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:274 +0xc5
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc8201781a0, 0xc820595ea0, 0x0, 0x0)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:234 +0xf21
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serveConn(0xc82035e9c0, 0x7f2944106000, 0xc8205a6840, 0x0, 0x0)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:198 +0x963
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve.func1(0xc82035e9c0, 0x7f2944186028, 0xc82011e140)
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:109 +0x90
created by github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve
        /home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:114 +0x20a
```

The non-debug version of `explain` works just fine.

--- a/sql/explain.go
+++ b/sql/explain.go
@@ -87,10 +87,6 @@ func markDebug(plan planNode, mode explainMode) (planNode, *roachpb.Error) {
 			t.filter = nil
 			t.render = nil
 			t.qvals = nil
-		} else if _, ok := t.table.node.(*scanNode); !ok {
-			// TODO(radu): We don't support debug mode for selects with no table or with a
-			// virtual table (subquery).
-			return t, nil
 		}
 		// Mark the from node as debug (and potentially replace it).
 		newNode, err := markDebug(t.table.node, mode)
@@ -113,6 +109,14 @@ func markDebug(plan planNode, mode explainMode) (planNode, *roachpb.Error) {
 		// Replace the group node with the node it wraps.
 		return markDebug(t.plan, mode)
 
+	case *emptyNode:
+		// emptyNode supports DebugValues without any explicit enablement.
+		return t, nil
+
+	case *valuesNode:
+		// valuesNode supports DebugValues without any explicit enablement.
+		return t, nil
+
 	default:
 		return nil, roachpb.NewErrorf("TODO(pmattis): unimplemented %T", plan)
 	}
@@ -133,11 +137,24 @@ func populateExplain(v *valuesNode, plan planNode, level int) {
 	}
 }
 
+type debugValueType int
+
+const (
+	// The debug values do not refer to a full result row.
+	debugValuePartial debugValueType = iota
+
+	// The debug values refer to a full result row but the row was filtered out.
+	debugValueFiltered
+
+	// The debug values refer to a full result row.
+	debugValueRow
+)
+
 // debugValues is a set of values used to implement EXPLAIN (DEBUG).
 type debugValues struct {
 	rowIdx int
 	key    string
-	value  parser.Datum
+	value  string
 	output debugValueType
 }
 
@@ -168,6 +185,11 @@ func (n *explainDebugNode) ExplainPlan() (name, description string, children []p
 func (n *explainDebugNode) Values() parser.DTuple {
 	vals := n.plan.DebugValues()
 
+	keyVal := parser.DNull
+	if vals.key != "" {
+		keyVal = parser.DString(vals.key)
+	}
+
 	// The "output" value is NULL for partial rows, or a DBool indicating if the row passed the
 	// filtering.
 	outputVal := parser.DNull
@@ -182,8 +204,8 @@ func (n *explainDebugNode) Values() parser.DTuple {
 
 	return parser.DTuple{
 		parser.DInt(vals.rowIdx),
-		parser.DString(vals.key),
-		vals.value,
+		keyVal,
+		parser.DString(vals.value),
 		outputVal,
 	}
 }

--- a/sql/plan.go
+++ b/sql/plan.go
@@ -259,19 +259,6 @@ func (p *planner) releaseLeases(db client.DB) {
 	}
 }
 
-type debugValueType int
-
-const (
-	// The debug values do not refer to a full result row.
-	debugValuePartial debugValueType = iota
-
-	// The debug values refer to a full result row but the row was filtered out.
-	debugValueFiltered
-
-	// The debug values refer to a full result row.
-	debugValueRow
-)
-
 // planNode defines the interface for executing a query or portion of a query.
 type planNode interface {
 	// Columns returns the column names and types . The length of the
@@ -316,14 +303,22 @@ type emptyNode struct {
 	results bool
 }
 
-func (*emptyNode) Columns() []ResultColumn  { return nil }
-func (*emptyNode) Ordering() orderingInfo   { return orderingInfo{} }
-func (*emptyNode) Values() parser.DTuple    { return nil }
-func (*emptyNode) DebugValues() debugValues { return debugValues{} }
-func (*emptyNode) PErr() *roachpb.Error     { return nil }
+func (*emptyNode) Columns() []ResultColumn { return nil }
+func (*emptyNode) Ordering() orderingInfo  { return orderingInfo{} }
+func (*emptyNode) Values() parser.DTuple   { return nil }
+func (*emptyNode) PErr() *roachpb.Error    { return nil }
 
 func (*emptyNode) ExplainPlan() (name, description string, children []planNode) {
 	return "empty", "-", nil
+}
+
+func (*emptyNode) DebugValues() debugValues {
+	return debugValues{
+		rowIdx: 0,
+		key:    "",
+		value:  parser.DNull.String(),
+		output: debugValueRow,
+	}
 }
 
 func (e *emptyNode) Next() bool {

--- a/sql/scan.go
+++ b/sql/scan.go
@@ -577,15 +577,15 @@ func (n *scanNode) explainDebug(endOfRow bool) {
 	n.debugVals.key = n.prettyKey()
 
 	if n.implicitVals != nil {
-		n.debugVals.value = parser.DString(prettyDatums(n.implicitVals))
+		n.debugVals.value = prettyDatums(n.implicitVals)
 	} else {
-		// This conversion to DString is odd. `n.explainValue` is already a
-		// `Datum`, but logic_test currently expects EXPLAIN DEBUG output
-		// to come out formatted using `encodeSQLString`. This is not
-		// consistent across all printing of strings in logic_test, though.
+		// We convert any datum to string, which will eventually be returned as a DString.
+		// This conversion to DString is odd. `n.explainValue` is already a `Datum`, but logic_test
+		// currently expects EXPLAIN DEBUG output to come out formatted using `encodeSQLString`.
+		// This is not consistent across all printing of strings in logic_test, though.
 		// TODO(tamird/pmattis): figure out a consistent story for string
 		// printing in logic_test.
-		n.debugVals.value = parser.DString(n.explainValue.String())
+		n.debugVals.value = n.explainValue.String()
 	}
 	if endOfRow {
 		n.debugVals.output = debugValueRow

--- a/sql/testdata/explain_debug
+++ b/sql/testdata/explain_debug
@@ -52,3 +52,32 @@ EXPLAIN (DEBUG) SELECT * FROM abc
 0  /abc/primary/1/'one'    NULL  true
 1  /abc/primary/2/'two'    NULL  true
 2  /abc/primary/3/'three'  NULL  true
+
+query ITTB
+EXPLAIN (DEBUG) SELECT 3
+----
+0 NULL NULL true
+
+query ITTB
+EXPLAIN (DEBUG) VALUES (1, 2, 3), (4, 5, 6)
+----
+0 0 (1, 2, 3) true
+1 1 (4, 5, 6) true
+
+query ITTB
+EXPLAIN (DEBUG) SELECT * FROM (VALUES (1, 2, 3), (4, 5, 6)) AS a;
+----
+0 0 (1, 2, 3) true
+1 1 (4, 5, 6) true
+
+query ITTB
+EXPLAIN (DEBUG) SELECT * FROM (SELECT * FROM abc) AS sub WHERE a = 2;
+----
+0  /abc/primary/1/'one'    NULL  false
+1  /abc/primary/2/'two'    NULL  true
+2  /abc/primary/3/'three'  NULL  false
+
+query ITTB
+EXPLAIN (DEBUG) SELECT * FROM (SELECT * FROM abc WHERE a = 2) AS sub;
+----
+0  /abc/primary/2/'two'    NULL  true

--- a/sql/values.go
+++ b/sql/values.go
@@ -98,7 +98,7 @@ func (n *valuesNode) DebugValues() debugValues {
 	return debugValues{
 		rowIdx: n.nextRow - 1,
 		key:    fmt.Sprintf("%d", n.nextRow-1),
-		value:  n.rows[n.nextRow-1],
+		value:  n.rows[n.nextRow-1].String(),
 		output: debugValueRow,
 	}
 }

